cmd/soroban-cli/src/signer.rs

```diff
@@ -0,0 +1,277 @@
+    SorobanAuthorizedFunction, SorobanCredentials, Transaction, TransactionEnvelope,
+    TransactionSignaturePayload, TransactionSignaturePayloadTaggedTransaction,
+    TransactionV1Envelope, Uint256, WriteXdr,
+};
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error(transparent)]
+    Xdr(#[from] xdr::Error),
+    #[error("Error signing transaction {address}")]
+    MissingSignerForAddress { address: String },
+}
+
+fn requires_auth(txn: &Transaction) -> Option<xdr::Operation> {
+    let [op @ Operation {
+        body: OperationBody::InvokeHostFunction(InvokeHostFunctionOp { auth, .. }),
+        ..
+    }] = txn.operations.as_slice()
+    else {
+        return None;
+    };
+    matches!(
+        auth.first().map(|x| &x.root_invocation.function),
+        Some(&SorobanAuthorizedFunction::ContractFn(_))
+    )
+    .then(move || op.clone())
+}
+
+/// A trait for signing Stellar transactions and Soroban authorization entries
+pub trait Stellar {
+    /// The type of the options that can be passed when creating a new signer
+    type Init;
+    /// Create a new signer with the given network passphrase and options
+    fn new(network_passphrase: &str, options: Option<Self::Init>) -> Self;
+
+    /// Get the network hash
+    fn network_hash(&self) -> xdr::Hash;
+
+    /// Sign a transaction hash with the given source account
+    /// # Errors
+    /// Returns an error if the source account is not found
+    fn sign_txn_hash(
+        &self,
+        txn: [u8; 32],
+        source_account: &stellar_strkey::Strkey,
+    ) -> Result<DecoratedSignature, Error>;
+
+    /// Sign a Soroban authorization entry with the given address
+    /// # Errors
+    /// Returns an error if the address is not found
+    fn sign_soroban_authorization_entry(
+        &self,
+        unsigned_entry: &SorobanAuthorizationEntry,
+        signature_expiration_ledger: u32,
+        address: &[u8; 32],
+    ) -> Result<SorobanAuthorizationEntry, Error>;
+
+    /// Sign a Stellar transaction with the given source account
+    /// This is a default implementation that signs the transaction hash and returns a decorated signature
+    /// # Errors
+    /// Returns an error if the source account is not found
+    fn sign_txn(
+        &self,
+        txn: Transaction,
+        source_account: &stellar_strkey::Strkey,
+    ) -> Result<TransactionEnvelope, Error> {
+        let signature_payload = TransactionSignaturePayload {
+            network_id: self.network_hash(),
+            tagged_transaction: TransactionSignaturePayloadTaggedTransaction::Tx(txn.clone()),
+        };
+        let hash = Sha256::digest(signature_payload.to_xdr(Limits::none())?).into();
+        let decorated_signature = self.sign_txn_hash(hash, source_account)?;
+        Ok(TransactionEnvelope::Tx(TransactionV1Envelope {
+            tx: txn,
+            signatures: vec![decorated_signature].try_into()?,
+        }))
+    }
```

Comment:
The sign function assumes the tx hasn't been wrapped in a fee bump transaction, but ideally it would support all Stellar txs.